processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -562,13 +598,30 @@ private File makeIndexFiles(
       );
       writer.open();
 
-      List<Indexed<String>> dimValueLookups = Lists.newArrayListWithCapacity(indexes.size());
+      List<Indexed<String>> dimValueLookups = Lists.newArrayListWithCapacity(indexes.size() + 1);
       DimValueConverter[] converters = new DimValueConverter[indexes.size()];
+      boolean dimHasValues = false;
+      boolean[] dimHasValuesByIndex = new boolean[indexes.size()];
+
       for (int i = 0; i < indexes.size(); i++) {
         Indexed<String> dimValues = indexes.get(i).getDimValueLookup(dimension);
         if (!isNullColumn(dimValues)) {
+          dimHasValues = true;
+          dimHasValuesByIndex[i] = true;
           dimValueLookups.add(dimValues);
           converters[i] = new DimValueConverter(dimValues);
+        } else {
+          dimHasValuesByIndex[i] = false;
+        }
+      }
+
+      // Ensure the empty str is always in the dictionary if column is not null across indexes
```

Comment:
Comment here about *why* we need it would be good (at least mention that the MMappedIndexRowIterable depends on it for being able to equate missing dims with empty strings)